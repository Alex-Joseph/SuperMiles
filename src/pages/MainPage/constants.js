export const actionTypes = {
  UPDATE_BALANCE: 'UPDATE_BALANCE',
  INSUFFICIENT_BALANCE: 'INSUFFICIENT_BALANCE',
};
