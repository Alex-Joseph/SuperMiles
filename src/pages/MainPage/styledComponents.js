import styled from 'styled-components';


const StyledMarketPlaceWrapper = styled.div`
  display: grid;
  place-content: center;
  padding: 24px 12px;

  h1 {
    text-align: center;
  }
`;

const StyledProductsWrapper = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;

  div {
    display: grid;
    justify-items: center;
  }
`;

export { StyledMarketPlaceWrapper, StyledProductsWrapper };
