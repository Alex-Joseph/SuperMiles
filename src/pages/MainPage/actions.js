import { actionTypes } from './constants';


export const postTransaction = (balance, product, history) => (dispatch) => {
  if (balance >= product.cost) {
    dispatch(updateBalance(balance - product.cost, product.desc));
    history.push('/receipt');
  } else {
    dispatch(insufficientBalance());
  }
};

export const updateBalance = (newBalance, desc) => ({
  type: actionTypes.UPDATE_BALANCE,
  balance: newBalance,
  transaction: { desc },
});

export const insufficientBalance = () => ({
  type: actionTypes.INSUFFICIENT_BALANCE,
});
