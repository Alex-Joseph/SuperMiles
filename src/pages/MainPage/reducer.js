import { fromJS } from 'immutable';
import { actionTypes } from './constants';


export const initialState = fromJS({
  balance: 50000,
  transaction: null,
});

export const walletReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case actionTypes.UPDATE_BALANCE:
      return state
        .set('balance', action.balance)
        .set('transaction', action.transaction);
    default:
      return state;
  }
};
