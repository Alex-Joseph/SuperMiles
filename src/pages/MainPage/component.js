import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Form, Header, Image, Button, Divider, Modal, Icon } from 'semantic-ui-react';
import { StyledMarketPlaceWrapper, StyledProductsWrapper } from './styledComponents';
import Navbar from '../../components/Navbar';
import AirBudsBlue from '../../images/air_buds_blue.jpg';
import AirBudsBlack from '../../images/air_buds_black.jpg';


const BLACK = 'black';
const BLUE = 'blue';

const priceMap = {
  'black': 40000,
  'blue': 30000,
};

class MainPage extends Component {
  componentDidMount() {
    if (!this.props.loggedIn) {
      this.props.history.push('/login');
    }
  };

  state = { cart: null };

  handleCart = (e, { value }) => this.setState({ cart: value });

  handleConfirm = () => this.setState({ modalOpen: true });

  handleCancel = () => this.setState({ modalOpen: false });

  handleTransaction = () => {
    const { postTransaction, balance, history } = this.props;
    const product = {
      cost: priceMap[this.state.cart],
      desc: this.state.cart,
    };
    postTransaction(balance, product, history);
  };

  render() {
    const { cart } = this.state;
    const { balance, logoutUser, history } = this.props;

    return (
      <Fragment>
        <Navbar title="SuperMiles" />
        <StyledMarketPlaceWrapper>
          <Header as='h1'>Balance: {balance}</Header>
          <StyledProductsWrapper>
            <div>
              <Image src={AirBudsBlue} size='medium' rounded />
              <Form.Radio
                label='Blue - 40000 points'
                value={BLUE}
                checked={cart === BLUE}
                onChange={this.handleCart}
              />
            </div>
            <div>
              <Image src={AirBudsBlack} size='medium' rounded />
              <Form.Radio
                label='Black - 30000 points'
                value={BLACK}
                checked={cart === BLACK}
                onChange={this.handleCart}
              />
            </div>
          </StyledProductsWrapper>
          <Divider horizontal></Divider>
          <Button.Group>
            <Button id="cancel" onClick={() => logoutUser(history)}>
              Cancel
            </Button>
            <Button.Or />
            <Button
              positive
              id="purchase"
              onClick={this.handleConfirm}
              disabled={!cart}
            >
              Purchase
            </Button>
          </Button.Group>
        </StyledMarketPlaceWrapper>
        <Modal open={this.state.modalOpen} basic size='small'>
          <Header icon='shopping cart' content='Confirm Order' />
          <Modal.Content>
            <p>
              Would you like to confirm your order for {priceMap[cart]} points?
            </p>
          </Modal.Content>
          <Modal.Actions>
            <Button
              basic
              inverted
              color='red'
              id='close'
              onClick={this.handleCancel}
            >
              <Icon name='remove' /> No
            </Button>
            <Button
              inverted
              color='green'
              id='confirm'
              onClick={this.handleTransaction}
            >
              <Icon name='checkmark' /> Yes
            </Button>
          </Modal.Actions>
        </Modal>
      </Fragment>
    )
  }
}

MainPage.propTypes = {
  loggedIn: PropTypes.bool.isRequired,
  balance: PropTypes.number.isRequired,
  logoutUser: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
  postTransaction: PropTypes.func.isRequired,
};

export default MainPage
