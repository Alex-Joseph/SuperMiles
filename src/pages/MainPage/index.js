import { connect } from 'react-redux';
import MainPage from './component';
import {
  mapStateToProps,
  mapDispatchToProps,
} from './redux';


export default connect(mapStateToProps, mapDispatchToProps)(MainPage);
