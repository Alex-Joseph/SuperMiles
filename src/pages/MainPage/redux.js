import { logoutUser } from '../LoginPage/actions';
import { postTransaction } from './actions';


export const mapStateToProps = state => ({
  loggedIn: state.toJS().auth.loggedIn,
  balance: state.toJS().wallet.balance,
  insufficientBalance: state.toJS().wallet.insufficientBalance,
});

export const mapDispatchToProps = {
  logoutUser,
  postTransaction,
};
