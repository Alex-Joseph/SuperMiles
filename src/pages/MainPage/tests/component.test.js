import React from 'react';
import { shallow } from 'enzyme';
import MainPage from '../component';


describe('MainPage', () => {
  const postTransaction = jest.fn();
  const logoutUser = jest.fn();
  const push = jest.fn();
  const history = {push};
  const balance = 20000;
  const BLUE = 'blue';

  it('renders correct snapshot', () => {
    const wrapper = shallow(
      <MainPage
        loggedIn={true}
        balance={balance}
        history={history}
        logoutUser={logoutUser}
        postTransaction={postTransaction}
      />,
    );

    expect(wrapper.find('#purchase').props().disabled).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('when state is updated the correct radio button is checked', () => {
    const wrapper = shallow(
      <MainPage
        loggedIn={true}
        balance={balance}
        history={history}
        logoutUser={logoutUser}
        postTransaction={postTransaction}
      />,
    );

    wrapper.setState({cart: BLUE});

    const radioButton = wrapper.find('FormRadio').first();

    expect(radioButton.props().checked).toBe(true);
    expect(radioButton.props().value).toBe(BLUE);
  });

  it('calls logoutUser when the Cancel button is pressed', () => {
    const wrapper = shallow(
      <MainPage
        loggedIn={true}
        balance={balance}
        history={history}
        logoutUser={logoutUser}
        postTransaction={postTransaction}
      />,
    );

    wrapper.find('#cancel').simulate('click');

    expect(logoutUser).toHaveBeenCalledWith(history);
  });

  it('the modal can open and close when a product is selected', () => {
    const wrapper = shallow(
      <MainPage
        loggedIn={true}
        balance={balance}
        history={history}
        logoutUser={logoutUser}
        postTransaction={postTransaction}
      />,
    );

    wrapper.setState({cart: BLUE});
    wrapper.find('#purchase').simulate('click');

    expect(wrapper.find('Modal').props().open).toBe(true);

    wrapper.find('#close').simulate('click');

    expect(wrapper.find('Modal').props().open).toBe(false);
  });

  it('calls postTransaction when a user purchases a product', () => {
    const wrapper = shallow(
      <MainPage
        loggedIn={true}
        balance={balance}
        history={history}
        logoutUser={logoutUser}
        postTransaction={postTransaction}
      />,
    );

    wrapper.setState({cart: BLUE});
    wrapper.find('#purchase').simulate('click');
    wrapper.find('#confirm').simulate('click');

    const expectedProduct = {cost: 30000, desc: BLUE};
    expect(postTransaction).toHaveBeenCalledWith(balance, expectedProduct, history);
  });
});
