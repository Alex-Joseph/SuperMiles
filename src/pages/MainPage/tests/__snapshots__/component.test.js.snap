// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage renders correct snapshot 1`] = `
<Fragment>
  <Navbar
    title="SuperMiles"
  />
  <styled.div>
    <Header
      as="h1"
    >
      Balance: 
      20000
    </Header>
    <styled.div>
      <div>
        <Image
          as="img"
          rounded={true}
          size="medium"
          src="air_buds_blue.jpg"
          ui={true}
        />
        <FormRadio
          as={[Function]}
          checked={false}
          control={[Function]}
          label="Blue - 40000 points"
          onChange={[Function]}
          value="blue"
        />
      </div>
      <div>
        <Image
          as="img"
          rounded={true}
          size="medium"
          src="air_buds_black.jpg"
          ui={true}
        />
        <FormRadio
          as={[Function]}
          checked={false}
          control={[Function]}
          label="Black - 30000 points"
          onChange={[Function]}
          value="black"
        />
      </div>
    </styled.div>
    <Divider
      horizontal={true}
    />
    <ButtonGroup>
      <Button
        as="button"
        id="cancel"
        onClick={[Function]}
      >
        Cancel
      </Button>
      <ButtonOr />
      <Button
        as="button"
        disabled={true}
        id="purchase"
        onClick={[Function]}
        positive={true}
      >
        Purchase
      </Button>
    </ButtonGroup>
  </styled.div>
  <Modal
    basic={true}
    centered={true}
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    size="small"
  >
    <Header
      content="Confirm Order"
      icon="shopping cart"
    />
    <ModalContent>
      <p>
        Would you like to confirm your order for 
         points?
      </p>
    </ModalContent>
    <ModalActions>
      <Button
        as="button"
        basic={true}
        color="red"
        id="close"
        inverted={true}
        onClick={[Function]}
      >
        <Icon
          as="i"
          name="remove"
        />
         No
      </Button>
      <Button
        as="button"
        color="green"
        id="confirm"
        inverted={true}
        onClick={[Function]}
      >
        <Icon
          as="i"
          name="checkmark"
        />
         Yes
      </Button>
    </ModalActions>
  </Modal>
</Fragment>
`;
